### ISSUE TITLE
Memory leak caused by identical classes being generated

### ISSUE THREAD

**Issue by asocaciu** [cglib/cglib#80]
We're having an issue where our spring-based JSF webapp throws an OOM after a few days (Spring 4.2.6 with cglib 3.2.2). I've tracked it down to the proxy classes that are generated for the Spring/JSF beans: instead of reusing the already generated cglib class definition, new ones are generated quite frequently. Since the classes themselves won't get garbage collected we run out of memory after a few days.

I've created a small example (mirroring the way spring uses cglib internally) to reproduce the issue:
```
while (true) {
            Enhancer enhancer = new Enhancer();
            enhancer.setSuperclass(SampleClass.class);
            enhancer.setCallbackFilter(new CallbackFilter() {
                @Override
                public int accept(Method method) {
                    return 0;
                }
                @Override
                public boolean equals(Object obj) {
                    return true;
                }

                @Override
                public int hashCode() {
                    return 0;
                }
            });
            enhancer.setInterfaces(new Class[]{SampleI.class});
            enhancer.setCallback(new InvocationHandler() {
                @Override
                public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
                    if (method.getDeclaringClass() != Object.class
                            && method.getReturnType() == String.class) {
                        System.out.println("Hello cglib!");
                        return null;
                    } else {
                        throw new RuntimeException("Do not know what to do.");
                    }
                }});
            SampleClass proxy = (SampleClass) enhancer.create();
            proxy.test("2");
            Thread.sleep(30);
        }
```
SampleClass is just: 
```
public class SampleClass implements SampleI {

    public String test(String s) {
        return s;
    }

}
```
After every GC run, the cached generated classes are no longer used, instead new ones are generated. See attached shot from memory profiler
![2016-06-16 11_51_00-cg - yourkit java profiler 2016 02-b38 - 64-bit](https://cloud.githubusercontent.com/assets/771915/16112709/e478987c-33b8-11e6-9f3c-8de1483c39cc.png)

I've tracked this down to the LoadingCache and the use of a CallbackFilter on the Enhancer. The CallbackFilter is kept via WeakReference and used as part of the map key in the cache. If the filter is not referenced from outside it will be GC'd, so the entry in the cache won't be found any more.

**Comment by vlsi** [comment]
So "weakly reachable CallbackFilters" and "classloader reuse" do not play together well.

**Comment by sameb** [comment]
@vsli I'm assuming this is related to the changes you made recently?  Do you know which commits wiuld cause this regression?  Can we fix?

At this point, I'm mostly inclined to revert recent changes and go back to a known table state.

**Comment by vlsi** [comment]
That boils down to: if we add a proxy.class -> enhancerKey strong reference, then the classloader would keep object class alive

**Comment by vlsi** [comment]
@sameb , this is caused by https://github.com/cglib/cglib/pull/54
Previously, cglib holded a strong reference to CallbackFilter, and the change was to introduce weakreference, so callbackFilters can be reclaimed.

I do not get what you mean by "recent changes"

**Comment by sameb** [comment]
Basically all changes since the 3.2.0 release seem prone to regressions. 

Are you able to fix things so the proxy class holds a strong ref to the enhancer key?

**Comment by vlsi** [comment]
>Basically all changes since the 3.2.0 release seem prone to regressions.

Basically, any change is prone to regression.

>Are you able to fix things so the proxy class holds a strong ref to the enhancer key?

I think that should work.
Alternative option is to revert "weakreference" fix and live with #54 kind of issue somehow.

**Comment by vlsi** [comment]
The reference to key is not sufficient, as weak reference is in-between the key and callbackfilter. So there should be a strong reference from proxy class to callbackfilter instead.

**Comment by sameb** [comment]
There were lots of changes leading to 3.2.0 that didn't cause regressions, but the ones after that seem to have triggered a flood of bugs. I'm not very convinced the benefits of the changes are worth the destabilization of the code. 

As far as attempting to fix, I do not think we should go with an approach that creates a memory leak. So reopening #54 doesn't seem like a great idea.

**Comment by vlsi** [comment]
JFYI: #54 exists in pre 3.2.0 as well.

**Comment by sameb** [comment]
Ah, that's unfortunate.  To to be fair, I prefer years-old bugs that folks are familiar with to new ones we don't fully understand.  Do you think it's possible to fix this & #54 together, or does one preclude the other?

**Comment by vlsi** [comment]
The sketch for solution of both #80 and #54 is in #81

**Comment by sameb** [comment]
@vlsi  The fix looks good, but I have a concern about patching things to make them work with the current setup:  It seems like this setup is prone to leaving unreachable instances in the LoadingCache. It's a bad idea to have a key whose equals/hashcode can change, and it seems like it's changable now if there's a WeakReference as part of the key.  Can that change so the key of the cache is guaranteed immutable?

**Comment by sameb** [comment]
(Alternately, could something sweep through the cache [perhaps on future writes?] and remove entries that are no longer useful?)

**Comment by vlsi** [comment]
> It's a bad idea to have a key whose equals/hashcode can change

It does not change over time. Hash code is cached at instantiation, and equals is just equals.

> Alternately, could something sweep through the cache [perhaps on future writes?] and remove entries that are no longer useful

That enhancement makes sense.

**Comment by vlsi** [comment]
> I have a concern about patching things to make them work

The implementation in #81 is just quick&dirty to check if that works and blows existing tests. From theoretical point of view, I think having a strong reference between proxy class and CalbackFilter makes sense.

**Comment by sameb** [comment]
The original issue states, "The CallbackFilter is kept via WeakReference and used as part of the map key in the cache." -- if the key of the map is using the CallbackFilter within a WeakReference as part of its equals method, then equals can change over time.  That leads to memory leaks where values are stuck in the cache and can no longer be referenced.  Mutable objects (such as WeakReferences) should not be part of keys of a hash map.  Only immutable objects should be keys.

**Comment by vlsi** [comment]
#80 states that "cglib fails to reuse proxy classes across GC".
That can (should) be solved with something like #81.

The issue of "thombstones" is another issue, and I suggest to move it to separate issue/PR for simplified review.

**Comment by sameb** [comment]
IIUC, the issue is caused by the fact that the code is using a mutable object as a key to a hash map.  Sure, #81 is a patch on top of things that can fix it... but it doesn't change the fact that the underlying code is leaky right now.  I'm bringing it up in hopes that there are ways of making it not leaky.

**Comment by vlsi** [comment]
@sameb , I disagree.
It has nothing to do with "mutable" as you call it object.

WeakCacheKey's equals and hashCode methods do not change semantics over time: https://github.com/cglib/cglib/blob/28aa2484f1c4a85ac61eb11cb477040f217b7b0e/cglib/src/main/java/net/sf/cglib/core/WeakCacheKey.java#L23-L30

In other words, WeakCacheKey is immutable from `hashCode`, `equals` point of view.

**Comment by sameb** [comment]
It looks that way on the surface, but unfortunately it isn't true. The class extends WeakReference and equals delegates to get(). The return value of get() can change over time, so that means it's mutable.

**Comment by vlsi** [comment]
@sameb , the only way "get" can change over time is to become null.
equals has `ours != null && theirs != null` check, so the fact that `get()` changes over time does not magically propagate to "WeakCacheKey.equals changes over time".

WeakCacheKey.equals will answer in exactly the same way no matter if the reference has been cleared or not.

**Comment by vlsi** [comment]
@sameb , to put it in another way: are the any technical objections from using `WeakCacheKey` as a `HashMap` key except your own preferences of "mutable objects should never be used as keys"?

JFYI:  `java.lang.String` is mutable. Should we dodge all usages of `java.lang.String` as a map key?

**Comment by sameb** [comment]
String isn't mutable, sorry.  It's very much immutable.  Let's stay focused on the problem at hand here.

For the reporter's case, it doesn't matter much that the key is weak because it's an anonymous class whose equals delegates to instance equality (e.g, ==).  However, imagine a CallbackFilter that defines an equals/hashcode, like the one in [Guice's AOP]( https://github.com/google/guice/blob/master/core/src/com/google/inject/internal/ProxyFactory.java#L209).  Now take the following sequence of steps:
  1) Enhancer created w/ CallbackFilter wrapped in WeakCacheKey.
  2) CallbackFilter goes out of scope, WeakCacheKey expires the reference.
  3) User creates another enhance with a CallbackFilter that should be .equals to the one from (1).
  4) WeakCacheKey can't .get() equality to the first one because it expired, a new item is created and the old entry in the cache is effectively unreachable.

This is my technical objection -- it's not based on personal preference, it's based on standard best practices and knowing what causes bugs.  See for example this [stack overflow question](http://stackoverflow.com/questions/7842049/are-mutable-hashmap-keys-a-dangerous-practice) or [this article](http://www.java-fries.com/2014/09/mutable-keys-in-hashmap-dangerous-practice/), or any of the other results for a [google search](https://www.google.com/search?q=mutable+key+in+hashmap+java).

In this case we may be able to workaround it by enforcing that the filter will never expire while the class/classloader is around.  Still, that's just a patch on top of it, and if we go that route, we should be careful that we are 100% certain it covers all scenarios.

**Comment by vlsi** [comment]
>2) CallbackFilter goes out of scope, WeakCacheKey expires the reference.

That is why I suggest to add a strong reference from proxy class to the callback.

The justification is as follows:
1) proxy class will keep CallbackFilter available for reuse (by making EnhancerKey in a good shape)
2.1) As long as either CallbackFilter or proxy class is reachable, the proxy could be reused
2.2) If both CallbackFilter and proxy class are not reachable, then it is safe to throw that cached entry away.

Re Guice AOP, its IndicesCallbackFilter has instanceof check: https://github.com/google/guice/blob/master/core/src/com/google/inject/internal/ProxyFactory.java#L227
If by any chance `WeakCacheKey<guice.IndicesCallbackFilter>` gets expired, then that means IndicesCallbackFilter.class is unreachable === IndicesCallbackFilter.class.getClassLoader() is unreachable === there is no way other Instance of IndicesCallbackFilter would return "true" at `o instanceof IndicesCallbackFilter` check ==> there is no way to reuse existing cache entry.
Is that clear?

Well, there might be cases when CallbackFilter.equals returns true based on `Random.value` or just always returns true, however that is not enough to keep proxy classes always reachable.

Basically, we can have two options:
1) Give up and agree that cglib always keeps strong references to all proxy classes (== it can always reuse proper classes). The drawback is it will leak memory.

2) Have ability to garbage collect proxy classes. I'm not sure explicit `free(java.lang.Class)` API is suitable here. We are for something automatic, right? Then, the most straightforward java way approach to garbage collect some class is make classloader unreachable. That is what everybody knows. That is why I suggest to tie proxy-class lifetime to its classloader lifetime. That is what "keep proxyClass => callbackfilter" fix (#81) is doing. That is why I consider it (having a strong reference there) to be a proper solution, not a monkey-patch.

There always can be a [Morse/FM-coded packets via wait/notify on an interned String.](https://twitter.com/shipilev/status/743126754974412800) kind of inter-classloader communication, but I do not think we should care for those cases much.

**Comment by vlsi** [comment]
PS. String is mutable, at least in OpenJDK. `String.hashCode` mutates String object.
PPS.  "Immutable objects are safe for key of HashMap" does not imply "mutable objects are always unsafe for keys".
PPPS. I do get why having an object with non-constant value of `hashCode/equals` plays bad with `HashMap`s.

**Comment by sameb** [comment]
A few things:

 * Setting the callback filter as a static var in the enhanced class sounds like a reasonable fix in this case, although I would recommend you heavily document why it's necessary..

 * For the callback filters such as the Guice AOP scenario -- the filter can be GC'd with cglib v3.2.2.  The callback being GC'd doesn't imply that the class isn't reachable.  If the proposed fix goes in, then your analysis seems correct: it will only be GC'd if the class isn't available at all.

 * String mutability:  hashCode being set lazily doesn't imply mutability.  That's just a lazily instantiated field.  There's extensive docs in hashCode() about why it's safe to do so, and hashCode isn't used in equals(..), so there's no impact to equality.

 * Mutable objects are, by their very nature, much more dangerous to use as keys of a map.  You have to take great care to ensure that the objects aren't lost.  Using immutable objects is always a better choice when possible.

End result: the proposed patch looks good to me.  Can you clean it up and add tests?

**Comment by vlsi** [comment]
> Can you clean it up and add tests?

The test is in there: https://github.com/cglib/cglib/pull/81/files#diff-22812a70b02a776341d358e300fcb427R281
Do you have some specific test in mind?

Documentation & cleanup parts are missing.

**Comment by sameb** [comment]
ah test wasn't there last i glanced at it (i think, at least).  documentation would be great, and if you're able to do the cleanup too that would be even better.  ping the PR when it's ready?

**Comment by vlsi** [comment]
sure

**Comment by lucaspate** [comment]
When do you anticipate a release for this? We experienced a perm gen leak in production when we upgraded to Spring 4.2.6, which I believe was caused by this issue. I created a simple, exaggerated example of my problem here: https://github.com/lucaspate/bugs/tree/master/SpringAopPermGenLeak . I also posed about it on the Spring Framework Jira here: https://jira.spring.io/browse/SPR-14385 .

**Comment by sameb** [comment]
i'd like to review+merge @vlsi's outstanding pull requests, then cut a release after that.  will do shortly.

**Comment by sameb** [comment]
Can you try with the latest code in master and confirm things are fixed for
you?

**Comment by lucaspate** [comment]
@sameb I can confirm that my memory leak is gone in master.

**Comment by sameb** [comment]
@lucaspate, thanks for confirming.  i've released cglib 3.2.4 to maven.  it may take a few hours for it to show up.
